File: firefox/src/java/org/openqa/selenium/firefox/internal/SocketLock.java
/*
Copyright 2007-2009 WebDriver committers
Copyright 2007-2009 Google Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package org.openqa.selenium.firefox.internal;

import org.openqa.selenium.WebDriverException;

import java.io.IOException;
import java.net.BindException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.concurrent.TimeUnit;

/**
 * Implements {@link Lock} via an implementation that uses a well-known server socket.
 * 
 * @author gregory.block@gmail.com (Gregory Block)
 */
public class SocketLock implements Lock {
  private static final long DELAY_BETWEEN_SOCKET_CHECKS = 100;
  
  private final int lockPort;
  private final Socket lockSocket;

  /**
   * Constructs a new SocketLock.  Attempts to lock the lock will attempt to acquire the
   * specified port number, and wait for it to become free.
   * 
   * @param lockPort the port number to lock
   */
  public SocketLock(int lockPort) {
    this.lockPort = lockPort;
    lockSocket = new Socket();
  }

  /**
   * @inheritDoc
   */
  public void lock(long timeoutInMillis) throws WebDriverException {
    InetSocketAddress address = new InetSocketAddress("localhost", lockPort);

    // Calculate the 'exit time' for our wait loop.
    long maxWait = System.currentTimeMillis() + timeoutInMillis;

    // Attempt to acquire the lock until something goes wrong or we run out of time.
    do {
      try {
        if (isLockFree(address))
          return;
        Thread.sleep(DELAY_BETWEEN_SOCKET_CHECKS);
      } catch (InterruptedException e) {
        throw new WebDriverException(e);
      } catch (IOException e) {
        throw new WebDriverException(e);
      }
    } while (System.currentTimeMillis() < maxWait);

    throw new WebDriverException(
        String.format("Unable to bind to locking port %d within %d ms", lockPort, timeoutInMillis));
  }

  /**
   * @inheritDoc
   */
  public void unlock() {
    try {
      if (lockSocket.isBound()) lockSocket.close();
    } catch (IOException e) {
      throw new WebDriverException(e);
    }
  }
  
  /**
   * Test to see if the lock is free.  Returns instantaneously.
   * 
   * @param address the address to attempt to bind to
   * @return true if the lock is locked; false if it is not
   * @throws IOException if something goes catastrophically wrong with the socket
   */
  private boolean isLockFree(InetSocketAddress address) throws IOException {
    try {
      lockSocket.bind(address);
      return true;
    } catch (BindException e) {
      return false;
    }
  }
}

File: firefox/src/csharp/webdriver-firefox/Internal/SocketLock.cs
﻿using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;

namespace OpenQA.Selenium.Firefox.Internal
{
    /// <summary>
    /// Provides a mutex-like lock on a socket.
    /// </summary>
    internal class SocketLock : ILock
    {
        #region Private members
        private static int delayBetweenSocketChecks = 100;

        private int lockPort;
        private Socket lockSocket; 
        #endregion

        #region Constructor
        /// <summary>
        /// Initializes a new instance of the <see cref="SocketLock"/> class.
        /// </summary>
        /// <param name="lockPort">Port to use to acquire the lock.</param>
        /// <remarks>The <see cref="SocketLock"/> class will attempt to acquire the
        /// specified port number, and wait for it to become free.</remarks>
        public SocketLock(int lockPort)
        {
            this.lockPort = lockPort;
            lockSocket = new Socket(AddressFamily.InterNetwork, SocketType.Raw, ProtocolType.Icmp);
        } 
        #endregion

        #region ILock Members
        /// <summary>
        /// Locks the mutex port.
        /// </summary>
        /// <param name="timeoutInMilliseconds">The amount of time (in milliseconds) to wait for 
        /// the mutex port to become available.</param>
        public void LockObject(long timeoutInMilliseconds)
        {
            IPHostEntry hostEntry = Dns.GetHostEntry("localhost");

            //Use the first IPv4 address that we find
            IPAddress ipAddress = IPAddress.Parse("127.0.0.1");
            foreach (IPAddress ip in hostEntry.AddressList)
            {
                if (ip.AddressFamily == AddressFamily.InterNetwork)
                {
                    ipAddress = ip;
                    break;
                }
            }

            IPEndPoint address = new IPEndPoint(ipAddress, lockPort);

            // Calculate the 'exit time' for our wait loop.
            DateTime maxWait = DateTime.Now.AddMilliseconds(timeoutInMilliseconds);

            // Attempt to acquire the lock until something goes wrong or we run out of time.
            do
            {
                try
                {
                    if (IsLockFree(address))
                    {
                        return;
                    }

                    Thread.Sleep(delayBetweenSocketChecks);
                }
                catch (ThreadInterruptedException e)
                {
                    throw new WebDriverException("the thread was interrupted", e);
                }
                catch (IOException e)
                {
                    throw new WebDriverException("An unexpected error occured", e);
                }
            }
            while (DateTime.Now < maxWait);

            throw new WebDriverException(string.Format(CultureInfo.InvariantCulture, "Unable to bind to locking port {0} within {1} ms", lockPort, timeoutInMilliseconds));
        }

        /// <summary>
        /// Unlocks the mutex port.
        /// </summary>
        public void UnlockObject()
        {
            try
            {
                lockSocket.Close();
            }
            catch (IOException e)
            {
                throw new WebDriverException("An error occured unlocking the object", e);
            }
        }
        #endregion

        #region IDisposable Members
        /// <summary>
        /// Releases all resources associated with this <see cref="SocketLock"/>
        /// </summary>
        public void Dispose()
        {
            if (lockSocket != null && lockSocket.Connected)
            {
                lockSocket.Close();
            }

            GC.SuppressFinalize(this);
        }
        #endregion

        #region Support methods
        private bool IsLockFree(IPEndPoint address)
        {
            try
            {
                lockSocket.Bind(address);
                return true;
            }
            catch (SocketException)
            {
                return false;
            }
        } 
        #endregion
    }
}
